crates/doc/src/inference.rs

```diff
@@ -899,6 +909,11 @@ impl Shape {
         let default = union_option(lhs.default, rhs.default);
         let secret = union_option(lhs.secret, rhs.secret);
 
+        // Union of annotations is actually an _intersection_, which yields only
+        // the annotations that are guaranteed to apply at a given location.
+        let mut annotations = lhs.annotations;
+        annotations.retain(|k, _| rhs.annotations.contains_key(k));
```

Comment:
nit: technically this should assert that the values are also equal.

Reply:
:+1: done